ADR040: Safe replacement of nodes (#696)
alphagov/gsp: docs/architecture/adr/ADR040-cluster-stability-node-replacement.md

The [AWS docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html) say this about the WaitOnResourceSignals parameter:
```
Specifies whether the Auto Scaling group waits on signals from new instances during an update. Use this property to ensure that instances have completed installing and configuring applications before the Auto Scaling group update proceeds. AWS CloudFormation suspends the update of an Auto Scaling group after new EC2 instances are launched into the group. AWS CloudFormation must receive a signal from each new instance within the specified PauseTime before continuing the update. To signal the Auto Scaling group, use the cfn-signal helper script or SignalResource API.
```

Sometimes new worker nodes don't appear ready in Kubernetes until after the previous worker node they're replacing has been terminated. This should allow us to resolve that and ensure that for every instance terminated e.g. through launch template changes, we have a replacement already in-place.

We could use one of these approaches:
* When a worker node starts we could have it schedule a Job, selecting itself to run on, that calls SignalResource.

> * We could have a DaemonSet that calls SignalResource and then sleeps infinitely

Reviewer: I don't like the idea of keeping a pod running using memory, IP addr, resources ... feels like this should be a Job or maybe even just part of the cloudinit/userdata stuff on boot?

Would also be nice if the logic for waiting for actually "all pods are running ok" rather than just "is up" ... not sure is that was implied or not

### Action 2

Remove this CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template. Mutually exclusive with action 2.

#### Action 2A

Manual cordoning, draining and termination of nodes

Pros:
* 

Cons:
* This would make it some rather irritating toil as an engineer would have to manually run the process.
* It might be possible to make a mistake here and cordon/drain too many nodes at once without termination so therefore no replacement.

#### Action 2B

A script that we run that cordons, drains and terminates nodes - triggered manually.

Pros:
* Would be able to choose exactly when we replace nodes to optimise for low load, no other critical operations ongoing, staff availability in the event of a problem, etc.

Cons:
* We'd have to notice the new launch template and trigger it manually.

#### Action 2C

A script that runs regularly that terminates nodes - this would ensure the cluster is eventually updated, just not immediately.

Pros:
* Has the added benefit of regularly killing off nodes regardless of launch template changes, meaning nodes don't stick around for too long.

Cons:

> * If this was a CronJob inside the cluster, it might terminate itself while running and leave us with an inconsistent cluster? But if it's a Lambda maybe not. Or if it's written to only kill one node and wait until the next run to kill another we might also be okay.

Reviewer 2: I think it's fine

* Would not immediately replace all the nodes, rollout would be slow.

> * If the new launch template is broken, the cluster would slowly break itself through terminating the remaining healthy nodes.

Reviewer 2: I misunderstood this. Resolving

### Action 2D

> Write an operator that manages the ASGs for the worker nodes that works with / replaces the cluster autoscaler. Mutually exclusive with action 3.

Reviewer 3: I haven't given this a great deal of thought. But, in my head, there's a launch template and a list of AZs within which to autoscale. The node operator is then responsible for launching and terminating the instances based on some woolly notion of "load" or some such (using ASGs or not).

Pros:
* Would be able to manage both node replacement for a new launch template and scale for load at the same time.

Cons:
* More code for us to write and maintain.
* Potentially re-inventing the wheel if we have to replace the cluster autoscaler.

### Action 3

A lifecycle hook in the ASG that tells it to cordon and drain nodes etc. when it terminates them.

Pros:
* This might have some benefit if we want to e.g. use spot instances in the future, as we'd be confident we can quickly cordon and drain a node in the event of a notification saying a current worker node will be terminated shortly.

Cons:

> * Nodes might still get terminated too quickly for the cluster to cope.

Reviewer 2: I don't think this is true? Lifecycle hooks allow you to pause the ASG activity until the node is safely drained. You send an event saying that you're ready for the ASG to continue and then it carries on.  If you take longer than an hour to drain though you might see issues:

> By default, the instance remains in a wait state for one hour, and then the Auto Scaling group continues the launch or terminate process (Pending:Proceed or Terminating:Proceed). If you need more time, you can restart the timeout period by recording a heartbeat. If you finish before the timeout period ends, you can complete the lifecycle action, which continues the launch or termination process.

### Action 4

Scale autoscaler down so it's not running while we replace nodes. Mutually exclusive with action 2D.

> This might be doable with ASG lifecycle hooks?

Reviewer 2: I'm not sure that lifecycle hooks are the right place to handle this - lifecycle hooks trigger on a per-instance-start-or-stop basis but we want to scale the autoscaler down before the whole node rollout and start it again once the rollout is complete

Reviewer 2: docs https://docs.aws.amazon.com/autoscaling/ec2/userguide/lifecycle-hooks.html

Reviewer 3: IMO this is just a `kubectl scale` command executed before and after the `cluster_state` resource is `put` in the `deploy` job (maybe with a `kubectl wait` if needed/possible).

Author: So do we mean to always `kubectl scale` down before running any terraform change (and then back up at the end), assuming a cluster exists?

Reviewer 3: In pseudocode, following discussion with @Krenair:

```
if eks-cluster exists:
    if cluster-autoscaler deployment exists:
        scale autoscaler deployment to 0 replicas

terraform apply

if cluster-autoscaler deployment exists:
    scale autoscaler deployment to 2 replicas
```

```

Sometimes new worker nodes don't appear ready in Kubernetes until after the previous worker node they're replacing has been terminated. This should allow us to resolve that and ensure that for every instance terminated e.g. through launch template changes, we have a replacement already in-place.

We could use one of these approaches:
* When a worker node starts we could have it schedule a Job, selecting itself to run on, that calls SignalResource.
* We could have a DaemonSet that calls SignalResource and then sleeps infinitely

Note that SignalResource callers will need EC2 instance ID as well as the ID of the CloudFormation stack that created the ASG it comes from.

Pros:
* AWS should mostly take care of the process for us, we'd just need some code to call SignalResource.
* The cluster should not encounter any health issues if new instances are ready to accept new pods before existing instances get terminated

Cons:
* Our Concourse job would probably time out while this process happens, but we can probably adjust/live with that.

### Action 2: Custom process for rolling nodes instead of CloudFormation UpdatePolicy

> Remove the CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template. Mutually exclusive with action 2.

Reviewer 3: ```suggestion
Remove the CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template. Mutually exclusive with action 1.
```

#### Action 2A: Manually roll nodes

Manual cordoning, draining and termination of nodes

Pros:

> * None?

Reviewer 3: Total control over each step in the process.

#### Action 2B: Script for rolling nodes, triggered manually

A script that we run that cordons, drains and terminates nodes - triggered manually.

Pros:
* Would be able to choose exactly when we replace nodes to optimise for low load, no other critical operations ongoing, staff availability in the event of a problem, etc.

Cons:
* We'd have to notice the new launch template and trigger it manually.

#### Action 2C: Regularly terminate nodes

A script that runs regularly that terminates nodes - this would ensure the cluster is eventually updated, just not immediately.

Pros:
* Has the added benefit of regularly killing off nodes regardless of launch template changes, meaning nodes don't stick around for too long.

Cons:
* If this was a CronJob inside the cluster, it might terminate itself while running and leave us with an inconsistent cluster? But if it's a Lambda maybe not. Or if it's written to only kill one node and wait until the next run to kill another we might also be okay.
* Would not immediately replace all the nodes, rollout would be slow.
* If the new launch template is broken, the cluster would slowly break itself through terminating the remaining healthy nodes.

> ### Action 2D: K8s node operator

Reviewer 2: ```suggestion
#### Action 2D: K8s node operator
```

# ADR040: Ensuring cluster stability while replacing nodes

## Status

Pending

## Context

SREs sometimes need to change various things about a cluster's worker nodes, and therefore replace all of the nodes:
* AMI (e.g. for an EKS update)
* instance type
* anything else in the launch template in some way (e.g. instance role)

> In addition to this, in our Terraform we get the latest AMI ID for our EKS version from AWS - this means that as soon as AWS releases a new AMI for our current EKS version, the next time the cluster's deployer step runs for Terraform it will also replace all the nodes.

Reviewer 3: ```suggestion

In addition to this, in our Terraform we get the latest AMI ID for our EKS version from AWS - this means that as soon as AWS releases a new AMI for our current EKS version, the next time the cluster's deployer step runs for Terraform it will also replace all the nodes.
```

> We have a CloudFormation UpdatePolicy on our worker node auto-scaling-groups that tells it to replace 1 node at a time in each ASG. This would be fine if we didn't have 3 ASGs with a small number of nodes in each - right now this practice terminates too many instances at once and results in the cluster being very unhappy.

Reviewer 3: ```suggestion
We have a CloudFormation UpdatePolicy on our worker node auto-scaling-groups that tells it to replace 1 node at a time in each ASG. This would be fine if we didn't have 3 ASGs with a small number of nodes in each - right now this practice terminates too many instances at once and results in the cluster becoming unstable. This in turn causes outages in the applications running in the cluster.
```

Remove this CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template.

#### A

Manual cordoning, draining and termination of nodes

Cons:
* this would make it some rather irritating toil as an engineer would have to manually run the process

#### B

A script that we run that cordons, drains and terminates nodes - triggered manually.

Cons:
* we'd have to notice the new launch template and trigger it manually

#### C

A script that runs regularly that terminates nodes - this would ensure the cluster is eventually updated, just not immediately.

Cons:
* If this was a CronJob inside the cluster, it might terminate itself while running and leave us with an inconsistent cluster? But if it's a Lambda maybe not. Or if it's written to only kill one node and wait until the next run to kill another we might also be okay.

### Action 2

A lifecycle hook in the ASG that tells it to drain nodes etc. when it terminates them. This might have some benefit if we want to e.g. use spot instances in the future?

Cons:
* nodes might still get terminated too quickly for the cluster to cope

### Action 3

> Scale autoscaler down so it's not running while we replace nodes

Reviewer 3: We should add that this is AWS' recommended approach in their documentation (and maybe link to it: https://docs.aws.amazon.com/eks/latest/userguide/update-stack.html)

Cons:
* Introduces some risk if there is a burst in pods wanting to schedule while we are replacing nodes

> ## Decision

Reviewer 3: ```suggestion
### Action 4

Write an operator that manages the ASGs for the worker nodes that works with / replaces the cluster autoscaler.

Cons:
* More code for us to write and maintain

## Decision
```
